### PR TITLE
feat: new types everywhere

### PR DESCRIPTION
**Pull request by 0xClandestine**

### CODE REVIEW

File: src/contracts/core/DelegationManager.sol [hunk 1]
```diff
@@ -380,8 +380,8 @@ contract DelegationManager is
     function increaseDelegatedShares(
         address staker,
         IStrategy strategy,
-        uint256 existingPrincipalShares,
-        uint256 addedShares
+        Shares existingPrincipalShares,
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:383]: rename to `existingShares`

File: src/contracts/core/DelegationManager.sol [hunk 2]
```diff
@@ -380,8 +380,8 @@ contract DelegationManager is
...
+        WithdrawableShares addedShares
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:384]: rename to `addedWithdrawableShares`

File: src/contracts/core/DelegationManager.sol [hunk 3]
```diff
@@ -414,7 +414,7 @@ contract DelegationManager is
     function decreaseDelegatedShares(
         address staker,
         IStrategy strategy,
-        uint256 removedShares
+        WithdrawableShares removedShares
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:417]: rename to `removedWithdrawableShares`

File: src/contracts/core/DelegationManager.sol [hunk 4]
```diff
@@ -516,7 +516,7 @@ contract DelegationManager is
                 operator: operator, 
                 staker: staker, 
                 strategy: strategies[i],
-                existingPrincipalShares: 0,
+                existingPrincipalShares: uint256(0).wrapShares(),
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:519]: rename parameter to `existingShares`

File: src/contracts/core/DelegationManager.sol [hunk 5]
```diff
@@ -553,7 +553,13 @@ contract DelegationManager is
 
         for (uint256 i = 0; i < withdrawal.strategies.length; i++) {
             IShareManager shareManager = _getShareManager(withdrawal.strategies[i]);
-            uint256 sharesToWithdraw = WithdrawableShares.unwrap(withdrawal.delegatedShares[i].toWithdrawableShares(totalMagnitudes[i]));
+            
+            // forgefmt: disable-next-line
+            Shares sharesToWithdraw = withdrawal.delegatedShares[i]
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:558]: these should be WithdrawableShares

File: src/contracts/core/DelegationManager.sol [hunk 6]
```diff
@@ -589,8 +595,8 @@ contract DelegationManager is
         address operator,
         address staker,
         IStrategy strategy,
-        uint256 existingPrincipalShares,
-        uint256 addedShares,
+        Shares existingPrincipalShares,
+        WithdrawableShares addedShares,
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:599]: rename both of these

File: src/contracts/core/DelegationManager.sol [hunk 7]
```diff
@@ -602,8 +608,13 @@ contract DelegationManager is
         });
 
         // based on total magnitude, update operators delegatedShares
-        uint256 delegatedShares = DelegatedShares.unwrap(WithdrawableShares.wrap(addedShares).toDelegatedShares(totalMagnitude));
-        operatorDelegatedShares[operator][strategy] += delegatedShares;
+        DelegatedShares delegatedShares = addedShares.toDelegatedShares(totalMagnitude);
+
+        // forgefmt: disable-next-line
+        operatorDelegatedShares[operator][strategy] =
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:614]: can we just store DelegatedShares?

File: src/contracts/core/DelegationManager.sol [hunk 8]
```diff
@@ -651,16 +666,16 @@ contract DelegationManager is
...
 
             // delegatedShares for staker to place into queueWithdrawal
-            delegatedSharesToWithdraw[i] = WithdrawableShares.wrap(sharesToWithdraw[i]).toDelegatedShares(totalMagnitudes[i]);
-            uint256 principalSharesToWithdraw = Shares.unwrap(
-                delegatedSharesToWithdraw[i].toShares(depositScalingFactors[staker][strategies[i]])
-            );
+
+            delegatedSharesToWithdraw[i] = sharesToWithdraw[i].toDelegatedShares(totalMagnitudes[i]);
+            
+            Shares principalSharesToWithdraw = delegatedSharesToWithdraw[i].toShares(depositScalingFactors[staker][strategies[i]]);
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:672]: principalShares

File: src/contracts/core/DelegationManager.sol [hunk 9]
```diff
@@ -637,7 +652,7 @@ contract DelegationManager is
         address staker,
         address operator,
         IStrategy[] memory strategies,
-        uint256[] memory sharesToWithdraw,
+        WithdrawableShares[] memory sharesToWithdraw,
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:655]: withdrawableSharesToWithdraw

File: src/contracts/core/DelegationManager.sol [hunk 10]
```diff
@@ -651,16 +666,16 @@ contract DelegationManager is
...
+            delegatedSharesToWithdraw[i] = sharesToWithdraw[i].toDelegatedShares(totalMagnitudes[i]);
+            
+            Shares principalSharesToWithdraw = delegatedSharesToWithdraw[i].toShares(depositScalingFactors[staker][strategies[i]]);
 
             // TODO: maybe have a getter to get shares for all strategies, like getDelegatableShares
             // check sharesToWithdraw is valid
             // but for inputted strategies
-            uint256 principalSharesWithdrawable = shareManager.stakerStrategyShares(staker, strategies[i]);
-            require(principalSharesToWithdraw <= principalSharesWithdrawable, WithdrawalExeedsMax());
+            Shares principalSharesWithdrawable = shareManager.stakerStrategyShares(staker, strategies[i]);
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:677]: sharesWithdrawable

File: src/contracts/core/DelegationManager.sol [hunk 11]
```diff
@@ -735,40 +750,16 @@ contract DelegationManager is
...
-            uint256 existingShares = WithdrawableShares.unwrap(
-                Shares.wrap(existingPrincipalShares).toDelegatedShares(depositScalingFactors[staker][strategy]).toWithdrawableShares(
-                    totalMagnitude
-                )
-            );
-            newDepositScalingFactor = (existingShares + addedShares) * WAD
-                / (existingPrincipalShares + addedShares) * WAD / totalMagnitude;
-        }
+        Shares existingPrincipalShares,
+        WithdrawableShares addedShares
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:754]: rename

File: src/contracts/core/DelegationManager.sol [hunk 12]
```diff
@@ -602,8 +608,13 @@ contract DelegationManager is
...
         // based on total magnitude, update operators delegatedShares
-        uint256 delegatedShares = DelegatedShares.unwrap(WithdrawableShares.wrap(addedShares).toDelegatedShares(totalMagnitude));
-        operatorDelegatedShares[operator][strategy] += delegatedShares;
+        DelegatedShares delegatedShares = addedShares.toDelegatedShares(totalMagnitude);
+
+        // forgefmt: disable-next-line
+        operatorDelegatedShares[operator][strategy] = 
+            operatorDelegatedShares[operator][strategy]
+                .add(delegatedShares.unwrap())
+                .wrapDelegated();
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:617]: why do we wrap here?

File: src/contracts/core/DelegationManager.sol [hunk 13]
```diff
@@ -863,22 +854,29 @@ contract DelegationManager is
...
+    ) public view returns (WithdrawableShares[] memory shares) {
         address operator = delegatedTo[staker];
         for (uint256 i = 0; i < strategies.length; ++i) {
             IShareManager shareManager = _getShareManager(strategies[i]);
             // TODO: batch call for strategyManager shares?
             // 1. read strategy deposit shares
-            shares[i] = shareManager.stakerStrategyShares(staker, strategies[i]);
+            
+            // forgefmt: disable-next-item
+            shares[i] = shareManager.stakerStrategyShares(staker, strategies[i])
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:865]: these should already be WithdrawableShares

File: src/contracts/core/DelegationManager.sol [hunk 14]
```diff
@@ -863,22 +854,29 @@ contract DelegationManager is
...
-                    Shares.wrap(shares[i]).toDelegatedShares(depositScalingFactors[staker][strategies[i]])
-                        .toWithdrawableShares(totalMagnitude)
-                );
+                
+                // forgefmt: disable-next-item
+                shares[i] = shares[i]
+                    .unwrap()
+                    .wrapShares()
+                    .toDelegatedShares(depositScalingFactors[staker][strategies[i]])
+                    .toWithdrawableShares(totalMagnitude);
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:879]: this is kind of confusing. probably better to not store the initial values in `shares[i]`

File: src/contracts/core/DelegationManager.sol [hunk 15]
```diff
@@ -735,40 +750,16 @@ contract DelegationManager is
...
-    ) internal returns (uint256) {
-        uint256 newDepositScalingFactor;
-
-        if (existingPrincipalShares == 0) {
-            // existing shares are 0, meaning no existing delegated shares. In this case, the new depositScalingFactor
-            // is re-initialized to
-            newDepositScalingFactor = WAD / totalMagnitude;
-        } else {
-            // TODO: OVERFLOWS: switch to muldiv everywhere
-            // since
```
Comment (gpsanant) [inline src/contracts/core/DelegationManager.sol:749]: can we keep the comment

File: src/contracts/core/DelegationManager.sol (hunk 9 above)
Comment (0xClandestine) [reply src/contracts/core/DelegationManager.sol:655]: doesn't "sharesToWithdraw" already kinda infer that they're withdrawable shares?

File: src/contracts/core/DelegationManager.sol (hunk 8 above)
Comment (0xClandestine) [reply src/contracts/core/DelegationManager.sol]: what mean here?

Comment (0xClandestine) [reply src/contracts/core/DelegationManager.sol]: delegatedSharesToWithdraw -> sharesToWithdraw?

File: src/contracts/core/DelegationManager.sol (hunk 12 above)
Comment (0xClandestine) [reply src/contracts/core/DelegationManager.sol:617]: operatorDelegatedShares mapping returns type DelegatedShares

File: src/contracts/core/DelegationManager.sol (hunk 8 above)
Comment (gpsanant) [reply src/contracts/core/DelegationManager.sol]: principalSharesToWithdraw -> sharesToWithdraw

there should be no mention of principal

File: src/contracts/core/DelegationManager.sol (hunk 9 above)
Comment (gpsanant) [reply src/contracts/core/DelegationManager.sol:655]: ambigious between shares and withdrawableShares

File: src/contracts/core/DelegationManager.sol (hunk 12 above)
Comment (gpsanant) [reply src/contracts/core/DelegationManager.sol:617]: shouldn't .add return DelegatedShares?

File: src/contracts/interfaces/IDelegationManager.sol [hunk 16]
```diff
@@ -139,7 +139,7 @@ interface IDelegationManager is ISignatureUtils {
         // Array of strategies that the QueuedWithdrawal contains
         IStrategy[] strategies;
         // Array containing the amount of shares for withdrawal in each Strategy in the `strategies` array
-        uint256[] shares;
+        WithdrawableShares[] shares;
```
Comment (gpsanant) [inline src/contracts/interfaces/IDelegationManager.sol:142]: these should be DelegatedShares

**Pull request merged**